include/mqtt/tls_implementation.hpp

```diff
@@ -0,0 +1,28 @@
+// Copyright Takatoshi Kondo 2018
+//
+// Distributed under the Boost Software License, Version 1.0.
+// (See accompanying file LICENSE_1_0.txt or copy at
+// http://www.boost.org/LICENSE_1_0.txt)
+
+#if !defined(MQTT_TLS_IMPLEMENTATION_HPP)
+#define MQTT_TLS_IMPLEMENTATION_HPP
+
+#if defined(MQTT_USE_TLS)
+#if defined(MQTT_USE_GNU_TLS)
+#include <boost/asio/gnutls.hpp>
+// The following is used in 'endpoint.hpp' and is not defined by default for Gnu TLS.
+# define ERR_GET_REASON(l)       (int)( (l)         & 0xFFFL)
```

Comment:
Hrmm...

I'd rather this not be a macro.

inline constexpr auto get_tls_error_reason(uint16_t val) { return val &0xFFFL; } ?

But why only 3 F's? That looks like a bug as well....

Reply:
I copied this over from here:
https://github.com/openssl/openssl/blob/OpenSSL_1_1_0-stable/include/openssl/err.h

Reply:
Not sure that's an OK thing to do. Different code licenses.

I'm utterly perplexed as to what the heck openssl is thinking with that kind of bit-fiddling.

But please provide a generic function that can be called, similar to what i prototyped for you above.

Then, if we're using OpenSSL, call the macro. Otherwise, do whatever it is that GnuTLS needs to happen.

Reply:
Moved the logic to a function in `tls_implementation.hpp`.
For GnuTLS we're simply checking whether the int value is equal to the enum value now.